packages/smithy-core/src/smithy_core/retries.py

```diff
@@ -480,3 +482,122 @@ def record_success(self, *, token: retries_interface.RetryToken) -> None:
 
     def __deepcopy__(self, memo: Any) -> "StandardRetryStrategy":
         return self
+
+
+class TokenBucket:
+    """
+    Implements the token bucket algorithm for rate limiting with configurable fill
+    rate and capacity. Tokens are added to the bucket at a fixed rate (fill_rate).
+    Requests must acquire tokens before proceeding, and will wait if insufficient
+    tokens are available.
```

Comment:
This focuses a lot on how we're using the TokenBucket rather than what the token bucket does. Something like this may be more accurate.

```suggestion
    TokenBucket provides a collection of arbitrary tokens while managing issuance
    and refilling over time. This is controlled by a fill rate that can be variably adjusted.
    When tokens aren't available, the bucket will enforce a delay before attempting to
    reacquire tokens until one is available or the defined timeout is reached.
```